### PR TITLE
Use new Windows signature APIs from Microsoft.Security.Extensions package

### PR DESCRIPTION
**Pull request by PaulHigin**
# PR Summary

Replaces previous Windows internal catalog signature APIs with public APIs from Microsoft.Security.Extensions package.

## PR Context

This allows us to use public APIs for Windows catalog signatures.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/security/Authenticode.cs [hunk 1]
```diff
@@ -289,147 +293,108 @@ internal static Signature GetSignature(string fileName, string fileContent)
...
+
+                default:
+                    System.Diagnostics.Debug.Fail("Signature type can only be None, Authenticode or Catalog.");
+                    break;
             }
-            catch (TypeLoadException)
+
+            signature.IsOSBinary = fileSigInfo.IsOSBinary;
+
+            if (signature.SignatureType == SignatureType.Catalog && !Signature.CatalogApiAvailable.HasValue)
```
Comment (anamnavi) [inline src/System.Management.Automation/security/Authenticode.cs:366]: Would you want the CatalogApiAvailable.HasValue property to be gotten for the `Signature signature` instance created on L336, or on the type?
```suggestion
            if (signature.SignatureType == SignatureType.Catalog && !signature.CatalogApiAvailable.HasValue)
```

File: src/System.Management.Automation/security/Authenticode.cs [hunk 2]
```diff
@@ -289,147 +293,108 @@ internal static Signature GetSignature(string fileName, string fileContent)
...
             {
-                // If we don't have WTGetSignatureInfo, don't return a Signature.
-                Signature.CatalogApiAvailable = false;
-                return null;
+                Signature.CatalogApiAvailable = fileSigInfo.State != SignatureState.Invalid;
```
Comment (anamnavi) [inline src/System.Management.Automation/security/Authenticode.cs:368]: It appears it is set on the type, not instance for Signature.

**Review by anamnavi — APPROVED** [review]
LGTM, just some questions for my own clarity.

**Review by TravisEz13 — APPROVED** [review]

File: src/System.Management.Automation/security/Authenticode.cs [hunk 3]
```diff
@@ -289,147 +293,108 @@ internal static Signature GetSignature(string fileName, string fileContent)
...
+                {
+                    fileSigInfo = FileSignatureInfo.GetFromFileStream(fileStream);
+                    System.Diagnostics.Debug.Assert(fileSigInfo is not null, "Returned FileSignatureInfo should never be null.");
+                }
+                catch (Exception)
                 {
-                    NativeMethods.SIGNATURE_INFO sigInfo = new NativeMethods.SIGNATURE_INFO();
-                    sigInfo.cbSize = (uint)Marshal.SizeOf(sigInfo);
+                    // For any API error, enable fallback to WinVerifyTrust APIs.
+                    Signature.CatalogApiAvailable = false;
```
[the remark this answers is not in the document]
Comment (PaulHigin) [reply src/System.Management.Automation/security/Authenticode.cs:327]: Yes, I don't like this but don't want to change it here and cause a regression.

File: src/System.Management.Automation/security/Authenticode.cs (hunk 2 above)
Comment (PaulHigin) [reply src/System.Management.Automation/security/Authenticode.cs:368]: No not a type.  This is a static variable reference.

File: test/powershell/engine/Security/FileSignature.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+Describe "Windows platform file signatures" -Tags 'Feature' {
+
+    It "Verifies Get-AuthenticodeSignature returns correct signature for catalog signed file" -Skip:(!$IsWindows) {
+
+        if ($null -eq $env:windir) {
+            throw "Expected Windows platform environment path variable '%windir%' not available."
+        }
+
+        $filePath = Join-Path -Path $env:windir -ChildPath 'System32\WindowsPowerShell\v1.0\Modules\PSDiagnostics\PSDiagnostics.psm1'
```
[the remark this answers is not in the document]
Comment (PaulHigin) [reply test/powershell/engine/Security/FileSignature.Tests.ps1]: It is included with Windows installations.

**Comment by TravisEz13** [comment]
@PowerShell/powershell-maintainers This should have a good amount of time to verify it is stable before we back port.
We should consider this for 7.2 for compliance reasons.  I think we can wait for 7.0 to EOL and not back port.

**Comment by adityapatwardhan** [comment]
This needs more bake time. Let's consider this after a couple of preview releases.

**Comment by adityapatwardhan** [comment]
Considering that this change is not required for compliance for 7.0.x or 7.2.x, we should not backport.

**Pull request merged**